--- libssp/ssp.c.orig	2013-12-03 10:01:47.446937584 +0100
+++ libssp/ssp.c	2013-12-03 10:02:55.883288751 +0100
@@ -136,7 +136,7 @@
 #ifdef HAVE_SYSLOG_H
   /* Only send the error to syslog if there was no tty available.  */
   else
-    syslog (LOG_CRIT, msg3);
+    syslog (LOG_CRIT, "%s", msg3);
 #endif /* HAVE_SYSLOG_H */
 
   /* Try very hard to exit.  Note that signals may be blocked preventing
