--- ghdl/Makefile.in.orig	2016-09-14 18:54:18.662169578 +0200
+++ ghdl/Makefile.in	2016-09-14 19:06:18.129066623 +0200
@@ -154,7 +154,7 @@
 	$(CP) -p $(srcdir)/dist/gcc/lang-specs.h $(gcc_vhdl_dir)
 
 libs.vhdl.gcc:
-	$(MAKE) GHDL=$(bindir)/ghdl$(EXEEXT) vhdl.libs.all libs.vhdl.standard
+	$(MAKE) GHDL=$(bindir)/ghdl$(EXEEXT) ANALYZE_OPTS="$(ANALYZE_OPTS)" vhdl.libs.all libs.vhdl.standard
 
 ghdllib: libs.vhdl.gcc libgrt.a
 
