diff -urN ghdl-0.25-orig/vhdl/grt/config/win32.c ghdl-0.25/vhdl/grt/config/win32.c
--- ghdl-0.25-orig/vhdl/grt/config/win32.c	2006-03-10 02:47:29.000000000 +0100
+++ ghdl-0.25/vhdl/grt/config/win32.c	2006-08-27 04:37:42.000000000 +0200
@@ -130,6 +130,23 @@
   return res;
 }
 
+#include <math.h>
+
+double acosh (double x)
+{
+  return log (x + sqrt (x*x - 1));
+}
+
+double asinh (double x)
+{
+  return log (x + sqrt (x*x + 1));
+}
+
+double atanh (double x)
+{
+  return log ((1 + x) / (1 - x)) / 2;
+}
+
 #ifndef WITH_GNAT_RUN_TIME
 void __gnat_raise_storage_error(void)
 {
