Index: sem_stmts.adb
===================================================================
--- vhdl/sem_stmts.adb	(revision 73)
+++ vhdl/sem_stmts.adb	(revision 89)
@@ -461,14 +462,18 @@
    begin
       Ok := True;
       -- Find the signal.
-      Target := Get_Target (Stmt);
-      Target := Sem_Expression (Target, Sig_Type);
-      if Target /= Null_Iir then
-         Set_Target (Stmt, Target);
-         Check_Target (Stmt, Target);
-         Sem_Types.Set_Type_Has_Signal (Get_Type (Target));
+      if Sig_Type = Null_Iir then
+         Ok := False;
       else
-         Ok := False;
+         Target := Get_Target (Stmt);
+         Target := Sem_Expression (Target, Sig_Type);
+         if Target /= Null_Iir then
+            Set_Target (Stmt, Target);
+            Check_Target (Stmt, Target);
+            Sem_Types.Set_Type_Has_Signal (Get_Type (Target));
+         else
+            Ok := False;
+         end if;
       end if;
 
       Expr := Get_Reject_Time_Expression (Stmt);
