--- gcc-4.3.4/gcc/config/i386/linux-unwind.h.siginfo	2006-10-26 21:31:09.000000000 +0200
+++ gcc-4.3.4/gcc/config/i386/linux-unwind.h	2012-07-25 21:06:35.687182719 +0200
@@ -137,9 +137,9 @@
     {
       struct rt_sigframe {
 	int sig;
-	struct siginfo *pinfo;
+	siginfo_t *pinfo;
 	void *puc;
-	struct siginfo info;
+	siginfo_t info;
 	struct ucontext uc;
       } *rt_ = context->cfa;
       /* The void * cast is necessary to avoid an aliasing warning.
--- gcc-4.3.4/gcc/config/xtensa/linux-unwind.h.siginfo	2008-01-18 02:37:53.000000000 +0100
+++ gcc-4.3.4/gcc/config/xtensa/linux-unwind.h	2012-07-25 21:06:42.587806680 +0200
@@ -66,7 +66,7 @@
   struct sigcontext *sc;
 
   struct rt_sigframe {
-    struct siginfo info;
+    siginfo_t info;
     struct ucontext uc;
   } *rt_;
 
--- gcc-4.3.4/gcc/config/alpha/linux-unwind.h.siginfo	2006-10-26 21:31:09.000000000 +0200
+++ gcc-4.3.4/gcc/config/alpha/linux-unwind.h	2012-07-25 21:06:35.686182628 +0200
@@ -52,7 +52,7 @@
   else if (pc[1] == 0x201f015f)	/* lda $0,NR_rt_sigreturn */
     {
       struct rt_sigframe {
-	struct siginfo info;
+	siginfo_t info;
 	struct ucontext uc;
       } *rt_ = context->cfa;
       sc = &rt_->uc.uc_mcontext;
--- gcc-4.3.4/gcc/config/bfin/linux-unwind.h.siginfo	2007-09-05 18:23:42.000000000 +0200
+++ gcc-4.3.4/gcc/config/bfin/linux-unwind.h	2012-07-25 21:06:35.686182628 +0200
@@ -52,10 +52,10 @@
     {
       struct rt_sigframe {
 	int sig;
-	struct siginfo *pinfo;
+	siginfo_t *pinfo;
 	void *puc;
 	char retcode[8];
-	struct siginfo info;
+	siginfo_t info;
 	struct ucontext uc;
       } *rt_ = context->cfa;
 
--- gcc-4.3.4/gcc/config/ia64/linux-unwind.h.siginfo	2005-06-25 03:22:41.000000000 +0200
+++ gcc-4.3.4/gcc/config/ia64/linux-unwind.h	2012-07-25 21:06:35.687182719 +0200
@@ -51,7 +51,7 @@
       struct sigframe {
 	char scratch[16];
 	unsigned long sig_number;
-	struct siginfo *info;
+	siginfo_t *info;
 	struct sigcontext *sc;
       } *frame_ = (struct sigframe *)context->psp;
       struct sigcontext *sc = frame_->sc;
@@ -130,7 +130,7 @@
       struct sigframe {
 	char scratch[16];
 	unsigned long sig_number;
-	struct siginfo *info;
+	siginfo_t *info;
 	struct sigcontext *sc;
       } *frame = (struct sigframe *)context->psp;
       struct sigcontext *sc = frame->sc;
--- gcc-4.3.4/gcc/config/mips/linux-unwind.h.siginfo	2007-11-15 06:33:28.000000000 +0100
+++ gcc-4.3.4/gcc/config/mips/linux-unwind.h	2012-07-25 21:06:35.687182719 +0200
@@ -79,7 +79,7 @@
       struct rt_sigframe {
 	u_int32_t ass[4];  /* Argument save space for o32.  */
 	u_int32_t trampoline[2];
-	struct siginfo info;
+	siginfo_t info;
 	_sig_ucontext_t uc;
       } *rt_ = context->cfa;
       sc = &rt_->uc.uc_mcontext;
--- gcc-4.3.4/gcc/config/pa/linux-unwind.h.siginfo	2007-09-12 01:16:01.000000000 +0200
+++ gcc-4.3.4/gcc/config/pa/linux-unwind.h	2012-07-25 21:06:35.688182809 +0200
@@ -66,7 +66,7 @@
   int i;
   struct sigcontext *sc;
   struct rt_sigframe {
-    struct siginfo info;
+    siginfo_t info;
     struct ucontext uc;
   } *frame;
 
