--- Makefile.in.orig	2015-03-10 14:43:16.232451425 +0100
+++ Makefile.in	2015-03-10 14:52:46.713449823 +0100
@@ -16,7 +16,7 @@
 #  Software Foundation, 59 Temple Place - Suite 330, Boston, MA
 #  02111-1307, USA.
 CC=@CC@
-build=@build@
+build=i386-redhat-linux
 srcdir=@srcdir@
 GNATMAKE=@GNATMAKE@
 prefix=@prefix@
@@ -40,7 +40,7 @@
 ADAC=$(CC)
 
 # Debug
-OPT_FLAGS=-g
+OPT_FLAGS=-g -m32
 ADA_FLAGS=-gnata
 
 # Optimize, do not forget to use MODE=--genfast for iirs.adb.
