--- src/ortho/llvm/Makefile.orig	2015-03-11 16:51:14.898005017 +0100
+++ src/ortho/llvm/Makefile	2015-03-11 16:51:32.250624009 +0100
@@ -10,7 +10,7 @@
 $(ortho_exec): $(ortho_srcdir)/llvm/ortho_llvm.ads force llvm-cbindings.o
 	gnatmake -o $@ -aI$(ortho_srcdir)/llvm -aI$(ortho_srcdir) \
 	$(GNAT_FLAGS) ortho_code_main -bargs -E \
-	-largs llvm-cbindings.o `$(LLVM_CONFIG) --ldflags --libs --system-libs` -lc++ #-static
+	-largs llvm-cbindings.o `$(LLVM_CONFIG) --ldflags --libs --system-libs` -lstdc++ #-static
 
 llvm-cbindings.o: $(ortho_srcdir)/llvm/llvm-cbindings.cpp
 	$(CXX) -c  -I`$(LLVM_CONFIG) --includedir --cflags` -o $@ $<
