--- src/vhdl/evaluation.adb.orig	2015-03-12 15:10:19.489526836 +0100
+++ src/vhdl/evaluation.adb	2015-03-12 15:22:13.476472857 +0100
@@ -2564,6 +2564,9 @@
             if Right < Left then
                --  Null range.
                return 0;
+            elsif Right = 9223372036854775807
+              and then Left = -9223372036854775808 then
+               return 0;
             else
                Res := Right - Left + 1;
             end if;
@@ -2571,6 +2574,9 @@
             if Left < Right then
                --  Null range
                return 0;
+            elsif Left = 9223372036854775807
+              and then Right = -9223372036854775808 then
+               return 0;
             else
                Res := Left - Right + 1;
             end if;
