--- dist/gcc/dist.sh.orig	2015-03-10 15:14:31.937714262 +0100
+++ dist/gcc/dist.sh	2015-03-10 15:33:51.077201080 +0100
@@ -48,7 +48,7 @@
 GTKWAVE_VERSION=3.3.50
 
 # GHDL version (extracted from version.ads)
-VERSION=`sed -n -e 's/.*GHDL \([0-9.a-z]*\) (.*/\1/p' ../../version.ads`
+VERSION=`sed -n -e 's/.*GHDL \([0-9.a-z]*\) (.*/\1/p' ../../src/version.ads`
 
 CWD=`pwd`
 
@@ -83,13 +83,14 @@
   mkdir $VHDLDIR/libraries/ieee2008
   mkdir $VHDLDIR/grt
   mkdir $VHDLDIR/grt/config
+  mkdir $VHDLDIR/grt/fst
 }
 
 # Build Makefile
 do_Makefile ()
 {
   sed -e "/^####libraries Makefile.inc/r ../../libraries/Makefile.inc" \
-      -e "/^####grt Makefile.inc/r ../grt/Makefile.inc" \
+      -e "/^####grt Makefile.inc/r ../../src/grt/Makefile.inc" \
      < Makefile.in > $VHDLDIR/Makefile.in
   cp Make-lang.in $VHDLDIR/Make-lang.in
 }
@@ -103,20 +104,26 @@
 lfiles="config-lang.in lang-options.h lang-specs.h"
 for i in $lfiles; do ln -sf $CWD/$i $VHDLDIR/$i; done
 
-for i in $cfiles; do ln -sf $CWD/../../$i $VHDLDIR/$i; done
+for i in $cfiles; do
+    if [ -f $CWD/../../src/$i ]; then
+	ln -sf $CWD/../../src/$i $VHDLDIR/$i
+    else
+	ln -sf $CWD/../../src/vhdl/$i $VHDLDIR/$i
+    fi
+done
 
 for i in ghdl.texi ghdl.1; do ln -sf $CWD/../../doc/$i $VHDLDIR/$i; done
 
-for i in $tfiles; do ln -sf $CWD/../$i $VHDLDIR/$i; done
+for i in $tfiles; do ln -sf $CWD/../../src/vhdl/translate/$i $VHDLDIR/$i; done
 
-for i in $ortho_files; do ln -sf $CWD/../../ortho/$i $VHDLDIR/$i; done
+for i in $ortho_files; do ln -sf $CWD/../../src/ortho/$i $VHDLDIR/$i; done
 
 for i in $ortho_gcc_files; do
-  ln -sf $CWD/../../ortho/gcc/$i $VHDLDIR/$i
+  ln -sf $CWD/../../src/ortho/gcc/$i $VHDLDIR/$i
 done
 
 for i in $ghdl_files; do
-  ln -sf $CWD/../ghdldrv/$i $VHDLDIR/ghdldrv/$i
+  ln -sf $CWD/../../src/ghdldrv/$i $VHDLDIR/ghdldrv/$i
 done
 
 for i in $libraries_files; do
@@ -124,15 +131,19 @@
 done
 
 for i in $grt_files; do
-    ln -sf $CWD/../grt/$i $VHDLDIR/grt/$i
+    ln -sf $CWD/../../src/grt/$i $VHDLDIR/grt/$i
 done
 
 for i in $grt_config_files; do
-    ln -sf $CWD/../grt/config/$i $VHDLDIR/grt/config/$i
+    ln -sf $CWD/../../src/grt/config/$i $VHDLDIR/grt/config/$i
+done
+
+for i in $grt_fst_files; do
+    ln -sf $CWD/../../src/grt/fst/$i $VHDLDIR/grt/fst/$i
 done
 
 for i in $psl_files; do
-    ln -sf $CWD/../../psl/$i $VHDLDIR/$i
+    ln -sf $CWD/../../src/psl/$i $VHDLDIR/$i
 done
 }
 
--- dist/gcc/dist-common.sh.orig	2015-03-10 17:10:26.630605013 +0100
+++ dist/gcc/dist-common.sh	2015-03-10 18:03:56.474542092 +0100
@@ -106,7 +106,37 @@
 trans_be.ads
 trans_be.adb
 trans_analyzes.ads
-trans_analyzes.adb"
+trans_analyzes.adb
+trans.adb
+trans.ads
+trans-chap1.adb
+trans-chap1.ads
+trans-chap2.adb
+trans-chap2.ads
+trans-chap3.adb
+trans-chap3.ads
+trans-chap4.adb
+trans-chap4.ads
+trans-chap5.adb
+trans-chap5.ads
+trans-chap6.adb
+trans-chap6.ads
+trans-chap7.adb
+trans-chap7.ads
+trans-chap8.adb
+trans-chap8.ads
+trans-chap9.adb
+trans-chap9.ads
+trans-chap12.adb
+trans-chap12.ads
+trans-chap14.adb
+trans-chap14.ads
+trans-foreach_non_composite.adb
+trans-foreach_non_composite.ads
+trans-helpers2.adb
+trans-helpers2.ads
+trans-rtis.adb
+trans-rtis.ads"
 
 ortho_files="
 ortho_front.ads"
@@ -203,9 +233,13 @@
 grt-astdio.adb
 grt-avhpi.adb
 grt-avhpi.ads
+grt-avhpi_utils.adb
+grt-avhpi_utils.ads
 grt-avls.ads
 grt-avls.adb
 grt-c.ads
+grt-change_generics.adb
+grt-change_generics.ads
 grt-disp.adb
 grt-disp.ads
 grt-disp_rti.adb
@@ -218,6 +252,9 @@
 grt-errors.ads
 grt-files.adb
 grt-files.ads
+grt-fst.adb
+grt-fst.ads
+grt-fst_api.ads
 grt-hooks.adb
 grt-hooks.ads
 grt-images.adb
@@ -304,6 +341,15 @@
 pthread.c
 win32.c"
 
+grt_fst_files="
+config.h
+fastlz.c
+fastlz.h
+fstapi.c
+fstapi.h
+lz4.c
+lz4.h"
+
 psl_files="
 psl.ads
 psl-build.adb
