--- vhdl/grt/grt.adc.noruntime	2005-11-08 21:01:08.000000000 +0100
+++ vhdl/grt/grt.adc	2008-01-06 17:17:53.000000000 +0100
@@ -34,17 +34,17 @@
 --
 --  This files is *not* names gnat.adc, in order to ease the possibility of
 --  not using it.
-pragma Restrictions (No_Exception_Handlers);
+--pragma Restrictions (No_Exception_Handlers);
 --pragma restrictions (No_Exceptions);
 pragma Restrictions (No_Secondary_Stack);
 --pragma Restrictions (No_Elaboration_Code);
 pragma Restrictions (No_Dispatch);
 pragma Restrictions (No_Io);
 pragma Restrictions (No_Finalization);
-pragma restrictions (no_dependence => Ada.Tags);
+--pragma restrictions (no_dependence => Ada.Tags);
 pragma restrictions (no_dependence => GNAT);
 pragma Restrictions (Max_Tasks => 0);
 pragma Restrictions (No_Task_Hierarchy);
 pragma Restrictions (No_Implicit_Heap_Allocations);
 pragma Suppress (Overflow_Check);
-pragma No_Run_Time;
+--pragma No_Run_Time;
