--- gcc/vhdl/grt/grt-cbinding.c.orig	2009-12-30 00:24:06.000000000 +0100
+++ gcc/vhdl/grt/grt-cbinding.c	2009-12-30 00:27:36.000000000 +0100
@@ -43,6 +43,12 @@
   fprintf (stream, "%g", val);
 }
 
+void
+__ghdl_fprintf_clock (FILE *stream, int a, int b)
+{
+  fprintf (stream, "%3d.%03d", a, b);
+}
+
 #if 1
 void
 __gnat_last_chance_handler (void)
--- gcc/vhdl/grt/grt-stats.adb.orig	2009-12-30 00:24:17.000000000 +0100
+++ gcc/vhdl/grt/grt-stats.adb	2009-12-30 00:29:20.000000000 +0100
@@ -71,10 +71,8 @@
 
    procedure Put (Stream : FILEs; Val : Clock_T)
    is
-      Fmt : constant String := "%3d.%03d" & Character'Val (0);
-
-      procedure fprintf (Stream : FILEs; Fmt : Address; A, B : Clock_T);
-      pragma Import (C, fprintf);
+      procedure Fprintf_Clock (Stream : FILEs; A, B : Clock_T);
+      pragma Import (C, Fprintf_Clock, "__ghdl_fprintf_clock");
 
       Sec : Clock_T;
       Ms : Clock_T;
@@ -84,7 +82,7 @@
       --  Avoid overflow.
       Ms := ((Val mod One_Second) * 1000) / One_Second;
 
-      fprintf (Stream, Fmt'Address, Sec, Ms);
+      Fprintf_Clock (Stream, Sec, Ms);
    end Put;
 
    procedure Put (Stream : FILEs; T : Time_Stats) is
