--- vhdl/Makefile.in.grtso	2009-09-20 04:09:28.000000000 +0200
+++ vhdl/Makefile.in	2009-12-14 16:13:46.000000000 +0100
@@ -462,6 +462,15 @@
 
 GRT_ELF_OPTS:=-Wl,--version-script=@/grt.ver -Wl,--export-dynamic
 
+GRT_ADD_FLAGS?=
+
+GRT_FLAGS:=$(GRT_FLAGS) $(GRT_ADD_FLAGS)
+
+GRT_SO_MAJOR?=1
+GRT_SO_MINOR?=0
+GRT_SO_PL?=0
+GNATVER?=4.5
+
 # Set target files.
 ifeq ($(filter-out i%86 linux,$(arch) $(osys)),)
   GRT_TARGET_OBJS=i386.o linux.o times.o
@@ -531,6 +540,14 @@
 	 run-bind.o main.o
 	$(GRT_RANLIB) $@
 
+grt-so: libghdlgrt$(soext).$(GRT_SO_MAJOR).$(GRT_SO_MINOR).$(GRT_SO_PL)
+
+libghdlgrt$(soext).$(GRT_SO_MAJOR).$(GRT_SO_MINOR).$(GRT_SO_PL): \
+	$(GRT_ADD_OBJS) run-bind.o main.o grt-files # grt-arch.ads
+	$(CC) -shared -Wl,-soname,libghdlgrt$(soext).$(GRT_SO_MAJOR) \
+	 $(GRT_FLAGS)  -Wl,-Bsymbolic -o $@ `sed -e "/^-/d" < grt-files` \
+	 $(GRT_ADD_OBJS) run-bind.o main.o -lgnat-$(GNATVER) -lc
+
 run-bind.adb: grt-force
 	gnatmake -c $(GNATFLAGS) -aI$(GRTSRCDIR) $(GRT_PRAGMA_FLAG) \
 	  ghdl_main $(GRT_ADAFLAGS) -cargs $(GRT_FLAGS)
--- vhdl/grt/config/amd64.S.orig	2005-10-31 22:56:04.000000000 +0100
+++ vhdl/grt/config/amd64.S	2009-12-14 18:22:24.000000000 +0100
@@ -62,7 +62,8 @@
 	mov	-16(%rbp), %rsi
 	mov	%rsi, -16(%rax)
 	/* The return function.  Must be 8 mod 16.  */
-	movq	$grt_stack_loop, -24(%rax)
+	leaq	grt_stack_loop(%rip), %rsi
+	movq	%rsi, -24(%rax)
 	/* The context.  */
 	mov	%rbp, -32(%rax)
 	mov	%rbx, -40(%rax)
